src/main/java/net/wurstclient/hacks/PortalEspHack.java

```diff
@@ -0,0 +1,272 @@
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.util.math.ChunkPos;
+import net.minecraft.world.chunk.Chunk;
+import net.minecraft.world.dimension.DimensionType;
+import net.wurstclient.Category;
+import net.wurstclient.events.CameraTransformViewBobbingListener;
+import net.wurstclient.events.PacketInputListener;
+import net.wurstclient.events.RenderListener;
+import net.wurstclient.events.UpdateListener;
+import net.wurstclient.hack.Hack;
+import net.wurstclient.hacks.portalesp.PortalEspBlockGroup;
+import net.wurstclient.hacks.portalesp.PortalEspRenderer;
+import net.wurstclient.settings.CheckboxSetting;
+import net.wurstclient.settings.ChunkAreaSetting;
+import net.wurstclient.settings.ColorSetting;
+import net.wurstclient.settings.EspStyleSetting;
+import net.wurstclient.util.ChunkSearcher;
+import net.wurstclient.util.ChunkSearcher.Result;
+import net.wurstclient.util.ChunkUtils;
+import net.wurstclient.util.MinPriorityThreadFactory;
+import net.wurstclient.util.RenderUtils;
+
+public final class PortalEspHack extends Hack implements UpdateListener,
+	PacketInputListener, CameraTransformViewBobbingListener, RenderListener
+{
+	private final EspStyleSetting style = new EspStyleSetting();
+	
+	private final PortalEspBlockGroup netherPortal =
+		new PortalEspBlockGroup(Blocks.NETHER_PORTAL,
+			new ColorSetting("Nether portal color",
+				"Nether portals will be highlighted in this color.", Color.RED),
+			new CheckboxSetting("Include nether portals", true));
+	
+	private final PortalEspBlockGroup endPortal =
+		new PortalEspBlockGroup(Blocks.END_PORTAL,
+			new ColorSetting("End portal color",
+				"End portals will be highlighted in this color.", Color.GREEN),
+			new CheckboxSetting("Include end portals", true));
+	
+	private final PortalEspBlockGroup endPortalFrame = new PortalEspBlockGroup(
+		Blocks.END_PORTAL_FRAME,
+		new ColorSetting("End portal frame color",
+			"End portal frames will be highlighted in this color.", Color.BLUE),
+		new CheckboxSetting("Include end portal frames", true));
+	
+	private final PortalEspBlockGroup endGateway = new PortalEspBlockGroup(
+		Blocks.END_GATEWAY,
+		new ColorSetting("End gateway color",
+			"End gateways will be highlighted in this color.", Color.YELLOW),
+		new CheckboxSetting("Include end gateways", true));
+	
+	private final List<PortalEspBlockGroup> groups =
+		Arrays.asList(netherPortal, endPortal, endPortalFrame, endGateway);
+	
+	private final ChunkAreaSetting area = new ChunkAreaSetting("Area",
+		"The area around the player to search in.\n"
+			+ "Higher values require a faster computer.");
+	
+	private final HashMap<ChunkPos, ChunkSearcher> searchers = new HashMap<>();
+	private final BiPredicate<BlockPos, BlockState> query =
+		(pos, state) -> state.getBlock() == Blocks.NETHER_PORTAL
+			|| state.getBlock() == Blocks.END_PORTAL
+			|| state.getBlock() == Blocks.END_PORTAL_FRAME
+			|| state.getBlock() == Blocks.END_GATEWAY;
```

Comment:
Not new, PortalESP has always worked this way. The old `currentBlockList` variable made it look as if the list could change when groups are enabled and disabled, but it couldn't. It was basically hardcoded from the start - I'm just making it obvious. 😅

I do think it makes sense to keep it this way, as restarting all of the ChunkSearchers every time a setting is changed can be very computationally expensive. And since there are few portals in a typical Minecraft world, the extra memory needed to store the locations of all four portal types is usually negligible.

Reply:
> One possible approach could be to maintain the results of the ChunkSearcher for all portal types, regardless of whether they are currently enabled or not. Then, when rendering, you can check the enabled status of each portal type and only render those that are enabled. This way, you avoid unnecessary computation when settings change, while still allowing for dynamic enabling/disabling of portal types.

Yep, that's exactly what the current code does. Now shut up. @nobotreply